### PR TITLE
feat: adding optional buttons and configuration to ModalInfoView

### PR DESCRIPTION
**Pull request by DylanRJ**
# DCMAW-6243: iOS | User must set an analytics preference

This PR implements a number of new protocols for configuring a ModalInfoView with additional complexity as optional.
These protocols provide ModalInfoView the ability to have buttons and configure other view elements.

# Checklist

## Before raising your pull request:
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with ticket ID and a short description about the feature or update
      i.e. _DCMAW-222: Added ReadID SDK to iOS app_
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed

- [x] Met all accessibility requirements?
    - [x] Checked dynamic type sizes are applied
    - [x] Checked VoiceOver can navigate your new code
    - [x] Checked a user can navigate only using a keyboard around your new code 

## Before merging your pull request:
- [x] Ensure that the code coverage and SonarCloud checks have passed
- [ ] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [x] Ran the app to ensure that no regressions have been caused by changes during code review.
- [x] Targeted the correct branch; `develop`, `release` or `main`

### CODE REVIEW

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewController.swift [hunk 1]
```diff
@@ -40,8 +41,49 @@ public final class ModalInfoViewController: BaseViewController, TitledViewContro
...
             }
-            bodyLabel.textColor = .gdsGrey
+            bodyLabel.textColor = (viewModel as? ModalInfoButtonsViewModel)?.bodyTextColour ?? .gdsGrey
             bodyLabel.accessibilityIdentifier = "bodyLabel"
         }
     }
+    
+    @IBOutlet private var primaryButton: RoundedButton! {
+        didSet {
+            if viewModel is ModalInfoButtonsViewModel {
```
Comment (benmckillop) [inline Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewController.swift:51]: I think you could combine these a little - the logic here would be if it is that `viewModel`, but there is no `buttonViewModel`, it wouldn't hide the button.

```suggestion
if let modalButtonsViewModel = viewModel as? ModalInfoButtonsViewModel,
   let buttonViewModel = modalButtonsViewModel.primaryButtonViewModel {
    primaryButton.setTitle(buttonViewModel.title, for: .normal)
    primaryButton.accessibilityIdentifier = "modal-info-primary-button"
} else {
    primaryButton.isHidden = true
}
```

Same then for the `secondaryButton`

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewModel.swift [hunk 2]
```diff
@@ -7,3 +7,15 @@ public protocol ModalInfoViewModel {
     var title: GDSLocalisedString { get }
     var body: GDSLocalisedString { get }
 }
+
+/// View model for the `ModalInfoViewController`
+/// - `bodyTextColour` type is ``UIColor?``
+/// - `primaryButtonViewModel` type is ``ButtonViewModel?``
+/// - `secondaryButtonViewModel` type is ``ButtonViewModel?``
+/// - `preventModalDismiss` type is ``Bool?``
+public protocol ModalInfoButtonsViewModel: ModalInfoViewModel {
```
Comment (benmckillop) [inline Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewModel.swift:16]: nice! :D

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewController.swift (hunk 1 above)
Comment (DylanRJ) [reply Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewController.swift:51]: Good shout 🙂

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewModel.swift [hunk 3]
```diff
@@ -7,3 +7,11 @@ public protocol ModalInfoViewModel {
     var title: GDSLocalisedString { get }
     var body: GDSLocalisedString { get }
 }
+
+/// View model for extra configuration on the `ModalInfoViewController`
+/// - `bodyTextColour` type is ``UIColor``
+/// - `preventModalDismiss` type is ``Bool``
+public protocol ModalInfoExtraViewModel {
+    var bodyTextColour: UIColor { get }
```
Comment (Oliver-Binns) [inline Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewModel.swift:15]: `bodyTextColour` possibly fits better as part of the `ModalInfoViewModel`, and then we can default it to whatever the default colour is (presumably `UIColor.label`?)

Also, please can we be consistent with Swift and use the American spelling (I know, I know 😣)

Comment (DylanRJ) [reply Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/ModalInfoView/ModalInfoViewModel.swift:15]: Cool cool, will do
Yeah I've never really known what to do with British vs American spelling in code but will follow that practice  from now on... time to adding Zs 😅

**Review by benmckillop — APPROVED** [review]

**Pull request merged**